### PR TITLE
[CodeGen] Remove checks for vectors in unsigned division prior to computing leading zeros

### PR DESCRIPTION
**Pull request by AZero13**
It turns out we can safely use DAG.computeKnownBits(N0).countMinLeadingZeros() with constant legal vectors, so remove the check for it.

### CODE REVIEW

**Review by arsenm — APPROVED** [review]

**Comment by topperc** [comment]
There's no early exit for vector in the TargetLowering code. There's only an early exit if the type isn't legal.

**Review by topperc — CHANGES_REQUESTED** [review]
I don't think this patch is NFC

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -5226,6 +5224,8 @@ MachineInstr *CombinerHelper::buildUDivUsingMul(MachineInstr &MI) {
     // TODO: Use undef values for divisor of 1.
     if (!Divisor.isOne()) {
 
+      unsigned KnownLeadingZeros =
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:5227]: This will get called for every element of a vector, but will always calculate the same value since it's using `LHS` which doesn't change. This is bad for compile time.

Either put it back where it was or count the leading zeros of `C`.

Comment (AtariDreams) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp]: Fixed!

**Comment by topperc** [comment]
This patch changes the generated code for this test for x86-64

```
define <4 x i32> @foo(<4 x i32> %x) {                                            
  %a = and <4 x i32> %x, <i32 255, i32 255, i32 255, i32 255>                    
  %b = udiv <4 x i32> %a, <i32 7, i32 7, i32 7, i32 7>                           
  ret <4 x i32> %b                                                               
} 
```

**Comment by topperc** [comment]
> It turns out we can safely use DAG.computeKnownBits(N0).countMinLeadingZeros() with constant legal vectors, so remove the check for it.

It is safe, but you have to write tests for it.

**Comment by AtariDreams** [comment]
> define <4 x i32> @foo(<4 x i32> %x) {

I actually added this check to avoid unforeseen bugs and turns out that was bad on my part. But yeah I will add the test now.

**Comment by topperc** [comment]
> > define <4 x i32> @foo(<4 x i32> %x) {
> 
> I actually added this check to avoid unforeseen bugs and turns out that was bad on my part. But yeah I will add the test now.

What check did you add? The original code in TargetLowering was written by me. https://reviews.llvm.org/D140750 I specifically didn't do vectors at the time because I hadn't thought through the math as I wrote

```
Wasn't sure if we need to apply the clip by divisor leading zeros uniformly across the whole vector when there are different divisors which would require another loop over the divisors. Or if we could treat each element individually and apply a clip inside BuildUDIVPattern.
```

**Review by topperc — CHANGES_REQUESTED** [review]
Add back request changes until tests are added.

**Comment by AtariDreams** [comment]
> Add back request changes until tests are added.



> > > define <4 x i32> @foo(<4 x i32> %x) {
> > 
> > 
> > I actually added this check to avoid unforeseen bugs and turns out that was bad on my part. But yeah I will add the test now.
> 
> What check did you add? The original code in TargetLowering was written by me. https://reviews.llvm.org/D140750 I specifically didn't do vectors at the time because I hadn't thought through the math. As I wrote at the time:
> 
> "Wasn't sure if we need to apply the clip by divisor leading zeros uniformly across the whole vector when there are different divisors which would require another loop over the divisors. Or if we could treat each element individually and apply a clip inside BuildUDIVPattern."

My bad, I was confusing selectiondag and globalisel

File: llvm/test/CodeGen/X86/combine-udiv.ll [hunk 2]
```diff
@@ -755,3 +755,64 @@ define <4 x i1> @boolvec_udiv(<4 x i1> %x, <4 x i1> %y) {
   %r = udiv <4 x i1> %x, %y
   ret <4 x i1> %r
 }
+
+define <4 x i32> @vector_udiv(<4 x i32> %x) {
```
Comment (topperc) [inline llvm/test/CodeGen/X86/combine-udiv.ll:759]: `vector_udiv` -> `vector_div_leading_zeros`?

**Comment by topperc** [comment]
PR title needs to be updated. As we've now shown, they were never redundant.

**Comment by AtariDreams** [comment]
> PR title needs to be updated. As we've now shown, they were never redundant.

Fixed!

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -5180,7 +5180,9 @@ MachineInstr *CombinerHelper::buildUDivUsingMul(MachineInstr &MI) {
   LLT ScalarShiftAmtTy = ShiftAmtTy.getScalarType();
 
   unsigned KnownLeadingZeros =
-      KB ? KB->getKnownBits(LHS).countMinLeadingZeros() : 0;
+      (!MI.getFlag(MachineInstr::MIFlag::IsExact) && KB)
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:5183]: This is not related to vectors and is not mentioned in the description. Separate patch?

Comment (AtariDreams) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp]: Fixed!

**Comment by AtariDreams** [comment]
@topperc This should be good to go now! Thank you!

**Comment by AtariDreams** [comment]
@topperc This should be good to go now! Thank you!

Please note I do not have commit perms...

**Review by topperc — APPROVED** [review]
LGTM

**Review by KanRobert — APPROVED** [review]
